infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -472,4 +507,57 @@ public static LogUnitServerConfig parse(Map<String, Object> opts) {
                     .build();
         }
     }
+
+    public static class LogUnitLock {
+        private final ReadWriteLock resetLock = new ReentrantReadWriteLock();
+
+        private void acquireReadLockAndExecuteSync(Runnable action) {
+
+            acquireReadLock()
+                    .thenApply(lock -> {
+                        action.run();
+                        return lock;
+                    })
+                    .whenComplete((lock, err) -> lock.unlock());
```

Comment:
I think this style is an overkill. It's kind of convoluted. If you still wan't to use this, then I suggest moving the code to the entry point, which is `processRequest`, it's more centralized that way.

Also, wouldn't this swallow the exception ? On master, at least it's logged with `uncaughtException`.

Another cleaner solution  to consider is to shutdown the LogUnit executor and execute the reset logic on the IO thread. After the reset logic completes you can install a new executor. That way in-flight operations will time out on transition. As opposed to be being evaluated with an older epoch.

Reply:
>Also, wouldn't this swallow the exception ?
No, because `exceptionally` already catches. any exceptions in CF

anyways, I deleted LogUnitLock

changed